backend/composer/services/cs_ingestion/neurondm_script.py

```diff
@@ -88,42 +85,142 @@ def for_composer(n, cull=False):
         log.warning(
             (n.id_, [[n.in_graph.namespace_manager.qname(e) for e in pos]
                      for pos in unaccounted_pos]))
-    return {k:v for k, v in fc.items() if v} if cull else fc
-
-
-def location_summary(neurons, services, anatent_simple=False):
-    import csv
-    OntTerm.query._services = services
-    locations = sorted(set(
-        OntTerm(pe.p) for n in neurons for pe in n.pes
-        if pe.e in n._location_predicates))
-    [_.fetch() for _ in locations]
-    def key(t):
-        return (t.prefix, t.label[0].lower()
-                if isinstance(t, tuple)
-                else t.lower())
-
-    if anatent_simple:
-        header = 'label', 'curie', 'iri'
-        rows = (
-            [header] +
-            [(_.label, _.curie, _.iri) for _ in sorted(locations, key=key)])
-        with open('/tmp/npo-nlp-apinat-location-summary.csv', 'wt') as f:
-            csv.writer(f, lineterminator='\n').writerows(rows)
+    return {k: v for k, v in fc.items() if v} if cull else fc
+
+
+def get_connections(n, lpes):
+    partial_order = n.partialOrder()
+    expected_origins = lpes(ilxtr.hasSomaLocatedIn)
+    expected_destinations = create_uri_type_dict(lpes, {ilxtr.hasAxonPresynapticElementIn: 'AXON-T',
+                                                        ilxtr.hasAxonSensorySubcellularElementIn: 'AFFERENT-T'})
+    expected_vias = create_uri_type_dict(lpes, {ilxtr.hasAxonLocatedIn: 'AXON', ilxtr.hasDendriteLocatedIn: 'DENDRITE'})
+
+    tmp_origins, tmp_vias, tmp_destinations = process_connections(partial_order,
+                                                                  set(expected_origins),
+                                                                  expected_vias,
+                                                                  expected_destinations
+                                                                  )
+
+    origins = merge_origins(tmp_origins)
+    vias = merge_vias(tmp_vias)
+    destinations = merge_destinations(tmp_destinations)
+    return origins, vias, destinations
+
+
+def create_uri_type_dict(lpes_func, predicate_type_map):
+    uri_type_dict = {}
+    for predicate, type_name in predicate_type_map.items():
+        for uri in lpes_func(predicate):
+            uri_type_dict[uri] = type_name
+    return uri_type_dict
+
+
+# FIXME: The following algorithm will behave incorrectly for cases:
+#  where an anatomical entity has more than one 'role' (Origin, Via, Destinaion)
+#  where we have connections that jump layers
```

Comment:
I can't improve these points until I see how neurondm models these edge cases @ddelpiano

Reply:
@afonsobspinto @ddelpiano in case we ingest parts of the neuromdm neurons I propose to only ingest the neurons that can be 100% ingested, if a neuron has dependencies or depends on other neurons that couldn't be ingested then we should not ingest the neuron

Reply:
I'm now doing a double validation. I check if the statements are valid on their own properties, and then I check if they are valid on their dependencies (the forward connections).
 
This way, if a valid on their own statement refers an invalid statement in their forward connection, the valid on their own statement will not be ingested. This goes against what Sue mentioned in the shared channel but I trust that this was mentioned to her posterior to that post of hers.

The second check relies on the idea that there are no partial ingestions.